File: app/src/main/java/com/robotnec/budget/app/dialogs/PickerDialog.java
package com.robotnec.budget.app.dialogs;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import android.widget.ArrayAdapter;

import com.robotnec.budget.core.domain.Entity;

import java.util.ArrayList;

/**
 * @author zak <zak@swingpulse.com>
 */
public abstract class PickerDialog<T extends Entity> extends DialogFragment {

    public static final String TAG = PickerDialog.class.getSimpleName();

    protected static final String BUNDLE_ITEMS = "BundleItems";

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final ArrayList<T> items = (ArrayList<T>) getArguments().getSerializable(BUNDLE_ITEMS);

        if (items == null) {
            throw new NullPointerException();
        }

        AlertDialog.Builder builderSingle = new AlertDialog.Builder(getContext());

        bindDialog(builderSingle);

        final ArrayAdapter<String> arrayAdapter = new ArrayAdapter<>(
                getContext(),
                getLayoutId());

        for (T item : items) {
            arrayAdapter.add(item.getName());
        }

        builderSingle.setNegativeButton(
                android.R.string.cancel,
                (dialog, which) -> dialog.dismiss());

        builderSingle.setAdapter(
                arrayAdapter,
                (dialog, which) -> PickerDialog.this.onClick(items.get(which)));
        return builderSingle.create();
    }

    protected abstract void bindDialog(AlertDialog.Builder builder);

    protected abstract int getLayoutId();

    protected abstract void onClick(T value);
}

File: app/src/main/java/com/robotnec/budget/app/dialogs/PickerDialog.kt
package com.robotnec.budget.app.dialogs

import android.app.Dialog
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v7.app.AlertDialog
import android.widget.ArrayAdapter

import com.robotnec.budget.core.domain.Entity

import java.util.ArrayList

/**
 * @author zak zak@swingpulse.com>
 */
abstract class PickerDialog<in T : Entity> : DialogFragment() {

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val items = arguments.getSerializable(BUNDLE_ITEMS) as ArrayList<T>

        val builderSingle = AlertDialog.Builder(context)

        bindDialog(builderSingle)

        val arrayAdapter = ArrayAdapter<String>(
                context,
                layoutId)

        for (item in items) {
            arrayAdapter.add(item.name)
        }

        builderSingle.setNegativeButton(
                android.R.string.cancel
        ) { dialog, _ -> dialog.dismiss() }

        builderSingle.setAdapter(
                arrayAdapter
        ) { _, which -> this@PickerDialog.onClick(items[which]) }
        return builderSingle.create()
    }

    protected abstract fun bindDialog(builder: AlertDialog.Builder)

    protected abstract val layoutId: Int

    protected abstract fun onClick(value: T)

    companion object {

        val TAG = PickerDialog::class.java.simpleName!!

        val BUNDLE_ITEMS = "BundleItems"
    }
}
